go/vt/tabletserver/txserializer/tx_serializer.go

```diff
@@ -0,0 +1,220 @@
+package txserializer
+
+import (
+	"container/list"
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/youtube/vitess/go/vt/vterrors"
+
+	vtrpcpb "github.com/youtube/vitess/go/vt/proto/vtrpc"
+)
+
+// TxSerializer can be used to serialize incoming transactions which target the
+// same row range i.e. table name and WHERE clause are identical.
+// Additional transactions are queued and woken up once it's their turn i.e.
+// they are the first entry in the queue.
+//
+// This implementation is a specialized version of the sync2.Consolidator class.
+// However, there are many substantial differences:
+// - Results are not shared between queued transactions.
+// - Only one waiting transaction and not all are notified when the current one
+//   has finished.
+// - Waiting transactions are woken up in FIFO order.
+// - Waiting transactions are unblocked if their context is done.
+type TxSerializer struct {
+	// Immutable fields.
+	limitGlobal      int
+	limitPerRowRange int
+
+	mu     sync.Mutex
+	queues map[string]*queue
```

Comment:
That doesn't mean we can start doing it ;-)

Overall, I don't think the list.List structure as FIFO is increasing the complexity. In my opinion, instead, it's the tight coupling of the TxSerializer and queue objects which makes the code ugly. But I don't really see a way out if we want to delete a queue object after all requests are done. (And I think we have to delete it because the number of different objects equals the number of accessed rows and is unbounded.) What do you think?